scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
+        if min_only:
+            _dijkstra_undirected_multi(indices,
+                                       csr_data, csgraph.indices,
+                                       csgraph.indptr,
+                                       csrT_data, csgraphT.indices,
+                                       csgraphT.indptr,
+                                       dist_matrix, predecessor_matrix,
+                                       source_matrix, limitf)
+        else:
+            _dijkstra_undirected(indices,
+                                 csr_data, csgraph.indices, csgraph.indptr,
+                                 csrT_data, csgraphT.indices, csgraphT.indptr,
+                                 dist_matrix, predecessor_matrix, limitf)
 
     if return_predecessors:
-        return (dist_matrix.reshape(return_shape),
-                predecessor_matrix.reshape(return_shape))
+        if min_only:
+            return (dist_matrix.reshape(return_shape),
+                    predecessor_matrix.reshape(return_shape),
+                    source_matrix.reshape(return_shape))
+        else:
+            return (dist_matrix.reshape(return_shape),
+                    predecessor_matrix.reshape(return_shape))
     else:
         return dist_matrix.reshape(return_shape)
 
 
-cdef _dijkstra_directed(
-            np.ndarray[ITYPE_t, ndim=1, mode='c'] source_indices,
-            np.ndarray[DTYPE_t, ndim=1, mode='c'] csr_weights,
-            np.ndarray[ITYPE_t, ndim=1, mode='c'] csr_indices,
-            np.ndarray[ITYPE_t, ndim=1, mode='c'] csr_indptr,
-            np.ndarray[DTYPE_t, ndim=2, mode='c'] dist_matrix,
-            np.ndarray[ITYPE_t, ndim=2, mode='c'] pred,
-            DTYPE_t limit):
-    cdef unsigned int Nind = dist_matrix.shape[0]
-    cdef unsigned int N = dist_matrix.shape[1]
-    cdef unsigned int i, k, j_source, j_current
-    cdef ITYPE_t j
+cdef _dijkstra_setup_heap_multi(FibonacciHeap *heap,
+                                FibonacciNode* nodes,
+                                int[:] source_indices,
+                                int[:] sources,
+                                double[:] dist_matrix,
+                                int return_pred):
+    cdef:
+        unsigned int Nind = source_indices.shape[0]
+        unsigned int N = dist_matrix.shape[0]
+        unsigned int i, k, j_source
+        FibonacciNode *current_node
 
-    cdef DTYPE_t next_val
+    for k in range(N):
+        initialize_node(&nodes[k], k)
 
-    cdef int return_pred = (pred.size > 0)
+    heap.min_node = NULL
+    for i in range(Nind):
+        j_source = source_indices[i]
+        dist_matrix[j_source] = 0
+        if return_pred:
+            sources[j_source] = j_source
+        current_node = &nodes[j_source]
+        current_node.state = SCANNED
+        current_node.source = j_source
+        insert_node(heap, &nodes[j_source])
+
+cdef _dijkstra_scan_heap_multi(FibonacciHeap *heap,
```

Comment:
same here if bounds checking can safely be deactivated

Reply:
done.